agent/src/main/java/com/teamscale/jacoco/agent/testimpact/CoverageToTeamscaleStrategy.java

```diff
@@ -0,0 +1,102 @@
+package com.teamscale.jacoco.agent.testimpact;
+
+import com.squareup.moshi.JsonAdapter;
+import com.squareup.moshi.Moshi;
+import com.teamscale.client.ClusteredTestDetails;
+import com.teamscale.client.EReportFormat;
+import com.teamscale.jacoco.agent.JacocoRuntimeController;
+import com.teamscale.jacoco.agent.options.AgentOptions;
+import com.teamscale.jacoco.agent.util.LoggingUtils;
+import com.teamscale.report.jacoco.dump.Dump;
+import com.teamscale.report.testwise.jacoco.JaCoCoTestwiseReportGenerator;
+import com.teamscale.report.testwise.jacoco.cache.CoverageGenerationException;
+import com.teamscale.report.testwise.model.TestExecution;
+import com.teamscale.report.testwise.model.TestwiseCoverage;
+import com.teamscale.report.testwise.model.TestwiseCoverageReport;
+import com.teamscale.report.testwise.model.builder.TestCoverageBuilder;
+import com.teamscale.report.testwise.model.builder.TestwiseCoverageReportBuilder;
+import org.slf4j.Logger;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
+import static java.util.stream.Collectors.toList;
+
+/**
+ * Strategy that records test-wise coverage and uploads the resulting report to Teamscale. Also handles the {@link
+ * #testRunStart(List, boolean, Long)} event by retrieving tests to run from Teamscale.
+ */
+public class CoverageToTeamscaleStrategy extends TestEventHandlerStrategyBase {
+
+	private final Logger logger = LoggingUtils.getLogger(this);
+
+	private final JsonAdapter<TestwiseCoverageReport> testwiseCoverageReportJsonAdapter = new Moshi.Builder().build()
+			.adapter(TestwiseCoverageReport.class);
+
+	private final TestwiseCoverage testwiseCoverage = new TestwiseCoverage();
+	private final List<TestExecution> testExecutions = new ArrayList<>();
+	private List<ClusteredTestDetails> availableTests = new ArrayList<>();
+	private final JaCoCoTestwiseReportGenerator reportGenerator;
+
+	public CoverageToTeamscaleStrategy(JacocoRuntimeController controller, AgentOptions agentOptions,
+									   JaCoCoTestwiseReportGenerator reportGenerator) {
+		super(agentOptions, controller);
+		this.reportGenerator = reportGenerator;
+
+		if (agentOptions.getTeamscaleServerOptions().commit == null) {
+			throw new UnsupportedOperationException("You must provide a commit via the agent's options." +
+					" Auto-detecting the git.properties does not work since we need the commit before any code" +
+					" has been profiled in order to obtain the prioritized test cases from the TIA.");
+		}
+	}
+
+	@Override
+	public String testRunStart(List<ClusteredTestDetails> availableTests, boolean includeNonImpactedTests,
+							   Long baseline) throws IOException {
+		this.availableTests = availableTests;
```

Comment:
I'm not sure I understand. With "setup for the first step" do you mean when the users want to record test-wise coverage for the first time for all their tests? In that case: how would TS know about any tests yet? Or am I misinterpreting what you meant?

Reply:
I mean if you are just about to try out TIA/integrating it into your build. As it might be very complicated to get the list of tests beforehand you could as a first step ignore the new tests/deleted tests scenarios and just collect the available tests as you go in /test/start /test/end.

Reply:
ah, that makes sense :-)

Reply:
I implemented that and adjusted the corresponding test